### DISCUSSION TITLE
GDScript: Add type hints on Callable

### DISCUSSION THREAD

**Discussion by dass2608** [godotengine/godot-proposals#9286] [Scripting]
I think most people agree that type hints can help to reduce runtime errors. GDScript type hints can uncover many errors early, before even running the code. Callables are also a useful feature, but sadly, they lack type hints.
I think it would be great, if Callables could have type annotations. As GDScript is inspired by Python, I would consider a syntax similar to Python:
```
var my_add_lambda: Callable[[int, int] int] = func(a, b):
    return a + b
```

This syntax could also be used in function parameters:
```
func my_function_with_callback(callback: Callable[[int, int], int]):
    ...
```

A annotation directly on the function definition might also make sense:
```
some_signal.connect(func(a:int, b:int) -> int: return a + b)
```

**Comment by azur-wolve** [comment]
using [] could lead to confusion with arrays
what about
```gdscript
var whatever: Callable((int, int) -> int)
```

**Reply by AThousandShips** [reply]
How would that be confusing? You can't use `float[]` in GDScript, so there's no confusion without the `Array[float]` part

**Reply by azur-wolve** [reply]
still can be confusing for users that come from other languages (C/C++, C#, Java, etc)
not showstopping though

using `()`isnt good idea either, easy confusion with explicit precedence

**Reply by AThousandShips** [reply]
Fair enough, but those are entirely unrelated and different languages so there shouldn't be any more confusion than, for example, that we don't use `{` for blocks etc.

**Reply by monokrome** [reply]
`Callable((int, int) -> int)` seems a lot more confusing to me. It looks like I'm assigning the result of a function to the type, while `Callable[]` syntax is already used to provide additional type information arrays and follows the semantics of Python. There's no ambiguity since array types would be `Array[type]` so it still denotes what type of thing it is.

**Reply by Shidoengie** [reply]
what about `func(int,int) -> int`?

**Reply by Charidan** [reply]
Why not `Callable[int, int, int]` and assume the final value is a return? No need to add extra characters.

This does lead to `Callable[int, void]` for a consumer function, in contrast to `Callable[void]` for a no-args no-return func, but this is fine.

**Comment by dalexeev** [comment]
The problem is that GDScript supports dynamic typing, so in general this requires runtime checks, similar to how typed arrays are implemented. Also, this would affect not only GDScript, but also the core. Godot doesn't have a unified type system, if you expect something like:

```gdscript
class Array[T]

func filter(method: Callable[[T], bool]) -> Array[T]
func map[TOther](method: Callable[[T], TOther]) -> Array[TOther]
func reduce[TAccum](method: Callable[[TAccum, T], TAccum], accum: TAccum|Nil = null) -> TAccum
```

**Comment by Shadowblitz16** [comment]
Can this be made into a proposal?
I just ran into an issue where I needed this.

**Reply by AThousandShips** [reply]
See:
* https://github.com/godotengine/godot-proposals/issues/10807
